docs/designs/high-level.md

```diff
@@ -1,33 +1,104 @@
+* Configuration Reconciler - In charge of OpenSearch configuration
+* TLS Reconciler - mapped to `security.tls`. In charge of TLS configuration (OpenSearch Security Plugin)
+
+
+Each Reconciler, takes the up-to-date Custom Resource, extracts the relevant data it is in charge of, and then checks whether something needs to be created or modified. For example, the Cluster Reconciler, creates a StatefulSet based on the pool defined in `nodePools`, retrieves the current StatefulSet and compares to see if it has changed, mandating deploying it to Kubernetes. Some differences require a series of steps to be taken prior to that. For example, if the number of replicas has decreased (say from 5 to 3), then assuming Graceful scale-down was defined, it calls the Scaler component to start the process of gracefully removing data from that pod, before decreasing the number of pods. Since some actions, especially removing shards away from an OpenSearch Node, takes time, a Reconciler needs to acknowledge a certain event was done before it can proceed to execute other steps.
+
+The Reconcilers and Components communicate via a Status object of the Custom Resource. It contains a data structure that exposes the current status of each reconciler and component.
+
+The Manager retrieves the current custom resource, based on the ID, from Kubernetes (using its API) and checks to see whether it was deleted. If it was, nothing needs to be done since all Kubernetes objects created by the operator are using [Owner Reference](https://kubernetes.io/blog/2021/05/14/using-finalizers-to-control-deletion/#owner-references) to the custom resource, thus are deleted in a cascading fashion.
+
+If it was created or updated, the Manager uses the retrieved custom resource YAML to parse it and produces a data structure representing it. It then executes a chain of Reconcilers, passing the Custom Resource data structure as main input.
+
+Each Reconciler executes and return two possible return values:
+1. Continue: I'm done, you can continue to the next Reconciler in the chain.
+2. Return-to-start: I've started an action, and changed the k8s objects / OpenSearch Configuration, let's go back and run the chain again to verify if some reconciler need to respond to that. When that value is returned, the Manager restarts the execution of the Reconciler chain
+
+On both accounts the reconciler updates its status in the Status element of the custom resource.
+
+Aside from Reconcilers we also have Components. Those are software objects whose responsibilities don't fall into the category of reconciling Kubernetes objects based on the Custom Resource. The components we have are:
+1. Scaler: in charge of doing graceful scale down of an OpenSearch Node.
+2. Auto-Scaler (Not written yet): Will respond to OpenSearch metrics polled periodically from Prometheus and modify amount Data Nodes and/or Coordinators of the cluster to respond to varying load.
+
```

Comment:
>so that an orderly teardown of the deployed opensearch cluster can be done.

@swoehrl-mw Can you explain how adding a finalizer leads to an orderly teardown?

Reply:
>If the Controller detects that a custom object is being deleted it will call all the 

I have that written down below, so maybe add it there ?

Reply:
>Unless there are specific reasons against it all Reconcilers should always add an OwnerReference to the objects they create that points to the `OpenSearchCluster` object. This way the objects will be deleted automatically by Kubernetes when the parent object is deleted.

This sounds more like a suggestion, but it seems that it should be an assumption, no? It should be phrased as such.

Reply:
> Can you explain how adding a finalizer leads to an orderly teardown?
With the finalizer in place the operator has to actively acknowledge that its teardown is complete (by removing the finalizer) so if needed the operator could do several reconcile runs for the teardown. Also (and I think this is the more important point) if the deletion of the custom resource happens while the operator is not online (restarting, crashed, whatever) we will not lose that delete event so we do not have any orphaned clusters.

> I have that written down below, so maybe add it there ?
Fine by me, I just think we need to have it somewhere in the document.

> This sounds more like a suggestion, but it seems that it should be an assumption, no? It should be phrased as such.
I meant it as a suggestion or rather default rule for all reconcilers. If you have a better wording, please.